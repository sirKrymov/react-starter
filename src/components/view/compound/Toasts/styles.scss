@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.toasts {
  list-style-type: none;
  z-index: 1000000;
  position: fixed;
  padding: 0 16px;
  bottom: 0;
  width: 300px;
  right: 0;

  &__item {
    background-color: var(--bgSecondary);
    border-radius: $border-radius;
    align-items: center;
    box-shadow: $box-shadow-card;
    transition: all 0.5s;
    max-height: 0;
    min-height: 60px;
    animation: fade-in 0.5s forwards;
    position: relative;
    display: flex;
    padding: 8px 20px;
    height: 100%;

    &--info {
      .toasts__item-type {
        background-color: var(--thirdy);
      }
    }

    &--warning {
      .toasts__item-type {
        background-color: var(--warning);
      }
    }

    &--success {
      .toasts__item-type {
        background-color: var(--valid);
      }
    }

    &--error {
      .toasts__item-type {
        background-color: var(--error);
      }
    }

    &--hide {
      animation: fade-out 0.3s forwards;
    }
  }

  &__item-type {
    border-bottom-left-radius: $border-radius;
    border-top-left-radius: $border-radius;
    justify-content: center;
    align-items: center;
    min-height: 60px;
    transform: translateY(-50%);
    position: absolute;
    display: flex;
    height: 100%;
    width: 20px;
    left: 0;
    top: 50%;
  }

  &__type-icon {
    color: var(--primary);
  }

  &__item-message {
    margin-left: 12px;
  }

  &__item-close-btn {
    position: absolute;
    right: 0;
    top: 0;
  }
}
