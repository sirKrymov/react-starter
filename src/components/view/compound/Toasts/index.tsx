// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactElement } from 'react';

import { createPortal } from 'react-dom';
import cn from 'classnames';
import {
  IoIosInformationCircleOutline,
  IoIosCheckmarkCircleOutline,
  IoIosCloseCircleOutline,
  IoIosWarning,
  IoIosClose
} from 'react-icons/io';

import { Typography } from 'components/view/atoms/Typography';
import { ReactIcon } from 'components/view/atoms/ReactIcon';
import { Loader } from 'components/view/atoms/loaders/Loader';

import './styles.scss';

import { Toast as ToastType } from 'types/toasts.type';

interface IProps {
  removeToast(id: ToastType['id']): void;
  toastsState: string;
  toasts: ToastType[];
}

export function Toasts({
  toastsState,
  toasts,
  removeToast
}: IProps): ReactElement | null {
  const toastsTypeIcons = {
    success: <IoIosCheckmarkCircleOutline />,
    warning: <IoIosWarning />,
    error: <IoIosCloseCircleOutline />,
    info: <IoIosInformationCircleOutline />
  };

  return toasts.length
    ? createPortal(
        <ul className="toasts">
          {toasts.map(({ isLoading, autohide, message, type, id }) => {
            const toastClass = cn({
              [`toasts__item`]: true,
              'toasts__item--hide': toastsState === `hide ${id}`,
              [`toasts__item--${type}`]: type
            });

            return (
              <li className={toastClass} key={id} id={id}>
                <div className="toasts__item-type">
                  <ReactIcon className="toasts__type-icon" size="md">
                    {toastsTypeIcons[type]}
                  </ReactIcon>
                </div>

                {isLoading && <Loader />}

                <Typography className="toasts__item-message" variant="body3">
                  {message}
                </Typography>

                {!autohide && (
                  <ReactIcon
                    className="toasts__item-close-btn"
                    pointer
                    onClick={(): void => removeToast(id)}
                    size="xl"
                  >
                    <IoIosClose />
                  </ReactIcon>
                )}
              </li>
            );
          })}
        </ul>,
        document.body
      )
    : null;
}
