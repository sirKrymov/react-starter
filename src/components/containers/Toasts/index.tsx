import React from 'react';

import { observer } from 'mobx-react-lite';

import { Toasts } from 'components/view/compound/Toasts';

import { useStores } from 'hooks/useStores';

export const ToastsContainer = observer(() => {
  const { toastsStore } = useStores();
  const { removeToast, toasts, state } = toastsStore;

  return toastsStore.toasts ? (
    <Toasts removeToast={removeToast} toasts={toasts} toastsState={state} />
  ) : null;
});
