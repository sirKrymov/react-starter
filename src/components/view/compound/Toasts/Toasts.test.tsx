import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { Toasts } from '../Toasts';

describe('<Toasts> compound component', () => {
  const handleClick = sinon.spy();

  let shallowComponent: ShallowWrapper;

  const createComponent = (): void => {
    shallowComponent = shallow(
      <Toasts removeToast={handleClick} toastsState="" toasts={[]} />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent();
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).toBeNull;
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });
  });
});
