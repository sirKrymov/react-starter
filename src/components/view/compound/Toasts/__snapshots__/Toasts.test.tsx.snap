// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toasts> compound component WHEN the component loads THEN it should match snapshot 1`] = `""`;
